Update on survival analysis in tidymodels (#539)
tidyverse/tidyverse.org: content/blog/survival-analysis-parsnip-adjacent/index.Rmd

Making survival analysis a first-class citizen in tidymodels requires touching several aspects across the collection of packages:

- Pre-processing which accommodates the data structure of censored data.
- A bigger selection of models which can be specified, fitted, and used for prediction.
- Additional performance metrics tailored to survival data.

Our focus so far has been on model specification, fitting, and prediction so let's dive in and see what's new!

### Model specification

Several engines can be used for multiple purposes with the same model type, e.g. the `rpart` engine for `decision_tree()` can be used for classification or regression. This distinction can be made in parsnip by specifying the mode of a model.
We have now introduced a new mode `"censored regression"` in parsnip for models which can be used for survival analysis. The aforementioned `decision_tree()` with the `rpart` engine can also be used to fit a survival decision tree:

```{r new-mode}
library(censored)

decision_tree() %>% 
  set_engine("rpart") %>% 
  set_mode("censored regression")

```

> You may notice a new package name here, `censored`. The number of models supported in tidymodels is growing. The core models live in parsnip, however there are several parsnip-adjacent extension packages with additional models such as [discrim](https://discrim.tidymodels.org/) for discriminant analysis models, [poissonreg](https://poissonreg.tidymodels.org/) for Poisson regression models, and [baguette](https://baguette.tidymodels.org/) for ensemble models via bagging. The [censored](https://tidymodels.github.io/censored/) package is the extension package for censored regression. 

Reviewer: Should this paragraph be restructured to move the link to {censored} further up?

### Model fitting

For fitting a model, parsnip typically supports a formula and a matrix interface, regardless of the interface of the engine. For censored regression models, we are currently only supporting a formula interface. Most engines already support this type of interface, and a formula allows us to specify the outcome via the standard `survival::Surv()` function, tying together event time and status. 

A formula also let's us specify stratification for a proportional hazards model. For the `survival::coxph()` function, which powers the `survival` engine for `proportional_hazards()`, this may look like 

<div align="center">
  `Surv(time, event) ~ x + z + strata(s)`.
</div>

> The glmnet package, which powers the `glmnet` engine, does not have a formula interface. It requires us to specify the strata by stratifying the response via `stratifySurv(y, s)`. The response is typically a `Surv` object, so you could end up with a formula like this:

Reviewer: The following sounds a little repetitive
 
> The glmnet package, which powers the `glmnet` engine,

Reviewer 2: ```suggestion
Sine other package can fit this model, we can try another; let's use the glmnet package. However, `glmnet::glmnet()`, does not have a formula interface and it requires us to specify the strata by stratifying the response via `stratifySurv(y, s)`. The response is typically a `Surv` object, so you could end up with a formula like this:
```

Reviewer 2: "sine" 🙄

<div align="center">
  `stratifySurv(Surv(time, event), strata) ~ x`.
</div>

> This works for fitting the model but the syntax is rather convoluted on the left-hand side of the formula. The specification of the stratification is inconsistent with the approach taken by the survival package. Additionally, without further modification, prediction fails because variables on the left-hand side of the formula are treated as response variables by parsnip and not available at prediction time.

Reviewer 3: ```suggestion
This works for fitting the model but the syntax is rather convoluted on the left-hand side of the formula. The specification of the stratification is inconsistent with the approach taken by the survival package. Additionally, without further modification, prediction fails because variables on the left-hand side of the formula are treated as response variables by parsnip and are not available at prediction time.
```

```{r strata}
library(survival)

bladder_train <- bladder[-c(1:3),]
bladder_test <- bladder[1:3,]

cox_spec <- proportional_hazards(penalty = 0.123) %>%
    set_engine("glmnet")

f_fit <- fit(cox_spec,
             Surv(stop, event) ~ rx + size + number + strata(enum),
             data = bladder_train)
```

### Prediction 

We have introduced several new prediction types: time to event (`time`), linear predictor (`linear_pred`), survival probability (`survival`), and hazard (`hazard`). Some of these types are time-dependent, e.g., the survival probability is the probability to survive beyond a certain time point. In keeping with the tidymodels design principle that each row of the data set you're predicting on should give you exactly one row in the dataset of predictions, `predict(type = "survival", time)` returns a nested tibble if `time` contains multiple time points.

```{r predict}
f_pred <- predict(f_fit, new_data = bladder_test,
                  type = "survival", time = c(5, 10, 15, 20))

# 3 rows for 3 observations

> f_pred 

Reviewer 2: I would use a bunch of time points, `unnest()` them, and show a ggplot of survival curves for a few samples. This would help demonstrate the utility of what we are doing.

Reviewer 3: +1 for showing how to unnest this structure

# 4 rows for 4 time points
f_pred$.pred[[1]]
```

> 

Reviewer 2: Maybe say something like "A good part of our work here is to enable the same level of feature across packages. For example..."

> Usually, the linear predictor is the linear combination of the predictors, weighted with the coefficients, often denoted as $x^{T} \beta$. This is what is being returned by the `glmnet` engine for `proportional_hazards()` models and by the `mboost` engine for `boost_tree()` models (e.g., with mode `"censored regression"`). For proportional hazards models, the linear predictor describes _relative_ risk, i.e., it refers to a pair of observations: the observation $x$ and a reference observation with value 0. The survival package by default centers the observation with the mean to allow for an interpretation of risk relative to a "standard" observation. For consistency with other engines, prediction of type `linear_pred` from `proportional_hazards()` models with engine `survival` also do not center the predictors.

Reviewer 3: ```suggestion
Usually, the linear predictor is the linear combination of the predictors, weighted with the coefficients, often denoted as $x^{T} \beta$. This is what is being returned by the `glmnet` engine for `proportional_hazards()` models and by the `mboost` engine for `boost_tree()` models (e.g., with mode `"censored regression"`). For proportional hazards models, the linear predictor describes _relative_ risk, i.e., it refers to a pair of observations: the observation $x$ and a reference observation with value 0. The survival package by default centers the observation with the mean to allow for an interpretation of risk relative to a "standard" observation. For consistency with other engines, predictions of type `linear_pred` from `proportional_hazards()` models with engine `survival` also do not center the predictors.
```

Reviewer 4: Wait, also does *not*? This sounds confusing to me; maybe take out the "also" to clarify that something different is happening?

> The prodlim package let's us simulate survival data with two covariates `X1` (binary) and `X2` (numeric) which both have an effect of $exp(1)$ on the hazard of the unobserved event time.

Reviewer 4: ```suggestion
The prodlim package lets us simulate survival data with two covariates `X1` (binary) and `X2` (numeric) which both have an effect of $exp(1)$ on the hazard of the unobserved event time.
```

```{r, simulate}
library(tidymodels)
library(ggplot2)
library(prodlim)
set.seed(43500)

# add 0.5 to X2 (with mean 0) to demonstrate the effect of centering
train_dat <- SimSurv(2000) %>% mutate(X2 = X2 + 0.5)
test_dat <- SimSurv(20) %>% mutate(X2 = X2 + 0.5)
test_pred <- test_dat[, 5:6]
```

We can compare the value of the linear predictor using the true coefficients of 1 with the predictions from the survival package. Plotting them against each other shows the offset of the mean of `X2`.

```{r lp-center}
fit_survival <- coxph(Surv(time, status) ~ X1 + X2, data = train_dat)
pred_survival <- predict(fit_survival, newdata = test_pred)

pred <- bind_cols(
  test_pred %>% mutate(linear_pred = X1 * 1  + X2 * 1),
  tibble(pred_survival = pred_survival)
)

mean(train_dat$X2)
```

```{r lp-center-plot}
pred %>% 
  ggplot(aes(linear_pred, pred_survival)) + 
  geom_point() +
  geom_abline(slope = 1, intercept = 0)
```

> As part of the modelling workflow, we may want to use these predictions with a performance metric such as concordance. This metric captures the probability that the prediction goes in the same direction as the actual data. While the data captures _survival_ time, the linear predictor for a proportional hazards model is a factor on the _risk_. The predicted survival time is longer if the linear predictor is lower. To make the prediction and the observation go in the same direction, we flip the sign of the prediction. This may be opposite to the sign of predictions obtained using the underlying engine directly.

Reviewer 2: We can justify the sign change by saying that our philosophy is that the calculation of performance metrics should be agnostic to ancillary information such as the model type or the training set. Changing the sign helps standardize the model assessment modules.

Reviewer 4: This part is still confusing to me 😔

```{r, lp-sign}
fit_censored <- proportional_hazards() %>% 
  set_engine("survival") %>% 
  fit(Surv(time, status) ~ X1 + X2, data = train_dat)
pred_censored <- predict(fit_censored, type = "linear_pred", new_data = test_pred)

pred <- bind_cols(
  pred,
  pred_censored %>% rename(pred_censored = .pred_linear_pred)
)

pred %>% 
  ggplot(aes(linear_pred, pred_censored)) + 
  geom_point() +
  geom_abline(slope = -1, intercept = 0)
```

>   

Reviewer 2: I would also allude to the idea of having a more complete method for specifying baseline hazards (as opposed to the "zero" and "centered" approaches. If you want to include an example, we could use a medical design with drug concentration (zero is control) and age. This would lend itself to a combination of zero and centering for the baseline.

Reviewer 2: Also, maybe show the same survival code example with a different model (like random forest).

```{r table, echo = FALSE, results = "asis"}
yep <- cli::symbol$tick
nope <- cli::symbol$cross
todo <- cli::symbol$circle

mod_names <- get_from_env("models")
model_info <-
  map_dfr(mod_names, ~ get_from_env(paste0(.x, "_predict")) %>% mutate(alias = .x))

done <- model_info %>%
  filter(mode == "censored regression") %>%
  select(model = alias, engine, type, mode) %>% 
  pivot_wider(names_from = type, 
              values_from = mode, 
              values_fill = nope, 
              values_fn = function(x) yep) %>% 
  mutate(linear_pred = case_when(model == "survival_reg" ~ todo,
                                 TRUE ~ linear_pred),
         time = case_when(time == nope ~ todo,
                          TRUE ~ time)
  )
to_do <- tribble(
  ~model, ~engine, ~survival, ~linear_pred, ~time, ~raw, ~quantile, ~hazard,
  "rand_forest", "ranger", todo, nope, todo, nope, nope, nope,
  "rand_forest", "randomForestSRC",  todo, nope, todo, nope, nope, nope
)

bind_rows(done, to_do) %>% 
  arrange(model, engine) %>% 
  knitr::kable(escape = FALSE) 
```

### What's next?

> We want survival analysis to be a first-class citizen in tidymodels which means having access to preprocessing, resampling, models, metrics, and tuning. The next steps on that path are a role for censoring indicator columns and a step `step_surv()` in recipes as well as a new ROC metric for survival models in yardstick. An adaption of workflows and tune is to follow after that. Additionally, we are in the process of figuring out how different types of censoring and time-dependent effects fit into our API.

Reviewer 2: I'd make the next steps into a bulleted list to make them visually stand out more. Maybe start GH issues and like to the issues so that people can provide direct feedback easily.

> We'd love to hear from you! Some questions we have include: Would you find this useful? Are there particular models/methods you'd like to see? Do you use time-dependent effects? Do you use interactions? Which types of censoring do you use?

Reviewer 2: Same as above. Make them easier to see and link to issues.

Reviewer 4: I think deciding on how to gather feedback is important here. The three main options we talked about in the meeting are:

- using GitHub (a single issue? invite to open an issue?)
- RStudio Community (could use our shortlink to send folks to the right tag/category)
- something like a Google Form that goes to Hannah's inbox

These all have positives/negatives. I think the most important things are that the call-to-action is clear and that Hannah is comfortable with managing it.

Author: It's going on RStudio Community!

date: 2021-09-28
author: Hannah Frick
description: >
    We are working on extending support for survival analysis in tidymodels. We
    are looking for early adopters to try out the new package called censored 
    and give feedback.

photo:
  url: https://unsplash.com/photos/n6vS3xlnsCc
  author: Kelley Bozarth

# one of: "deep-dive", "learn", "package", "programming", or "other"
categories: [deep-dive] 

> tags: [tidymodels, parnsip]

Reviewer 4: ```suggestion
tags: [tidymodels, parsnip]
```

Author: thanks!
